distributed/tests/test_asyncprocess.py

```diff
@@ -297,3 +300,120 @@ def test_terminate_after_stop():
     yield proc.start()
     yield gen.sleep(0.1)
     yield proc.terminate()
+
+
+def _worker_process(worker_ready, child_pipe):
+    # child_pipe is the write-side of the children_alive pipe held by the
+    # test process. When this _worker_process exits, this file descriptor should
+    # have no references remaining anywhere and be closed by the kernel. The
+    # test will therefore be able to tell that this process has exited by
+    # reading children_alive.
+
+    # Signal to parent process that this process has started and made it this
+    # far. This should cause the parent to exit rapidly after this statement.
+    worker_ready.close()
+
+    # The parent exiting should cause this process to os._exit from a monitor
+    # thread. This sleep should never return.
+    shorter_timeout = 2.5 # timeout shorter than that in the spawning this.
+    sleep(shorter_timeout)
+
+    # Unreachable if functioning correctly.
+    child_pipe.send("child should have exited by now")
+
+
+def _parent_process(child_pipe):
+    """ Simulate starting an AsyncProcess and then dying.
+
+    The child_alive pipe is held open for as long as the child is alive, and can
+    be used to determine if it exited correctly. """
+    def parent_process_coroutine():
+        await_worker_start, worker_ready = mp_context.Pipe(duplex=False)
+
+        # Need to use "spawn" context so that the child_pipe file descriptor can
+        # be passed to the parent onwards. This does not occur with the
+        # forkserver multiprocessing context, somehow new file descriptors are
+        # created.
+        worker = AsyncProcess(target=_worker_process,
+                              args=(worker_ready, child_pipe))
+
+        yield worker.start()
+
+        # Close our reference to the worker_ready pipe, so that the child has
+        # the only one.
+        worker_ready.close()
+
+        # Wait for the child process to have started.
+        yield PipeIOStream(await_worker_start.fileno()).read_until_close()
```

Comment:
`PipeIOStream` will probably not work on Windows. We will probably have to use a blocking call instead, such as `await_worker_start.read()`. But then the Pipe trick is not really useful anymore, and a `multiprocessing.Event` object may work just as well (and make the code a bit simpler).